taichi/backends/opengl/opengl_api.cpp

```diff
@@ -325,14 +318,25 @@ bool is_opengl_api_available() {
 
 void initialize_opengl(){TI_NOT_IMPLEMENTED}
 
-GLProgram *compile_glsl_program(std::string source) {
+CompiledGLSL::CompiledGLSL(std::string source)
+    : glsl(std::make_unique<GLProgram>()) {
```

Comment:
I think it's fine not to initialize `glsl`, since opengl is not supported in this case anyway.

Reply:
I want to but there'll be a compile error about `unique_ptr copy constructor implicitly deleted` if I do so...

Reply:
Hmm, strange..

Reply:
I think this is because you only defined `GLProgram` inside `#ifdef TI_WITH_OPENGL`, maybe add a dummy `struct GLProgram {};` in the `#else` branch?